MAINT: Use vectorcall for call forwarding in methods

seberg opened the pull request:
Mainly, I think this is a bit simpler, it also is a a fair bit faster when keyword arguments are being used.  (not that it matters most of the time)


---

Really, I mainly think this is a bit nicer, I ran a few benchmarks (not all) and this dropped out, which is probably true:
```
       before           after         ratio
     [4e69d516]       [e4bb12da]
     <main>           <forward-call-vectorcall>
-        2.21±0μs      2.10±0.02μs     0.95  bench_core.CountNonzero.time_count_nonzero_axis(2, 100, <class 'numpy.int32'>)
-     2.08±0.03μs      1.96±0.01μs     0.95  bench_core.CountNonzero.time_count_nonzero_multi_axis(1, 100, <class 'bool'>)
-        759±30ns          718±2ns     0.95  bench_core.StatsMethods.time_sum('intp', 1)
-     2.13±0.02μs      1.97±0.02μs     0.92  bench_core.CountNonzero.time_count_nonzero_multi_axis(3, 100, <class 'bool'>)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```
if you have lots of kwargs on a `sum` or so, its quite noticable, of course for most calls it isn't a whole lot of a difference.

seberg committed:
MAINT: Use vectorcall for call forwarding in methods

Mainly, I think this is a bit simpler, it also is a a fair bit faster
when keyword arguments are being used.  (not that it matters most of the
time)

seberg commented:
(main reason was that I thought I would do some changes to `dump` as a numpy 2.0 C-API smoke screen... and that lead me to doing this first.

mhvk reviewed (approved):
Very much in favour! And looks good with perhaps a thought in-line about `PY_VECTORCALL_ARGUMENTS_OFFSET`

seberg commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -63,34 +63,35 @@ NpyArg_ParseKeywords(PyObject *keys, const char *format, char **kwlist, ...)
-    int i, n;
+    /* 
+     * The only thing we need to do is add `self` to args, for that allocate
+     * a new one and copy it.  `PY_VECTORCALL_ARGUMENTS_OFFSET` or a stack
```
Yeah, I had it first, but then tried and `arr.sum()` doesn't set the flag, so I figured I leave out the code duplication.  I guess e.g. Cython might set it though.  So very happy to re-add it, also happy to do the heap optimization, but I suspect these are usually so heavy wight that it isn't noticable.

mhvk replied:
Since the routine is called by all of the methods, it seems a nice place to actually do the optimizations...

seberg replied:
Was about to just add it, but now tweaked the comment.  I think what is going on is that someone promises that it is OK to change the original `self` being passed (yay).

I wonder if Python will dare changing it, I think I never even added it to any of our method calls.  It seems the flag is always lost on the way, but I am not sure its intentional.

(Not that I mind Python adding it at all, the question would just be about the stable API.)

seberg replied:
Oh well... added the heap version, and just made it so that it will blow up if python ever starts passing it (because that is currently true for many places anyway).

I would be happy if python starts flagging it here though, may open an issue for it (or if Python just defines that this is always valid for methods (even if never flagged).

The new commit may be a bit ugly, so happy to tweak or revert, OTOH, its just one place.

seberg committed:
MAINT: Forward with stack allocation and fail for vectorcall argoffset

PY_VECTORCALL_ARGUMENTS_OFFSET seems never set, if it is the malloc
is bound to fail: After that we should add a path for it :).

mhvk commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -63,34 +63,48 @@ NpyArg_ParseKeywords(PyObject *keys, const char *format, char **kwlist, ...)
-    return ret;
+    new_args[0] = self;
+    memcpy(&new_args[1], args, original_arg_size);
+    PyObject *res = PyObject_Vectorcall(callable, new_args, len_args+1, kwnames);
```
For a moment, I wondered if there was a point to now set the `PY_VECTORCALL_ARGUMENTS_OFFSET` flag to indicate `self` can be overwritten, but I guess there is little point since this just goes out to python anyway.

Anyway, happy to just go with this!

seberg replied:
Here you would also indicate that `new_args[-1]` can be, so I don't think its valid.

mhvk replied:
I think if one were to use it, one would have to pass in `new_args+1`. But to be honest, I'm still not sure that I really understand how it is supposed to work.

seberg replied:
:shrug:, for `PyObject_VectorcallMethod()` the flag is lost along way internal to Python when calling an unbound method (it could be restored in the end in theory, but is not and the information was dropped).

I would think it does its job somewhere (`tp_vectorcall` does get it) and I am sure _bound_ methods use it to insert the `self` (but then don't forward it here either, even though they could).

It would make sense for us getting the flag almost always here, but just flipping the switch may break the stable API that got away with forgetting about it, so not sure its worth pursuing...

seberg replied:
Anyway, I think it makes quite a lot of sense. The one likely thing here we could potentially ask from Python:  It looks like `args[-1]` is *always* `self` (which isn't surprising).

Frankly, as long as PyPy does that too, we could just abuse it, but it would maybe be better to ask for it to be documented to be true.

seberg committed:
TST: Add minimal test to cover many-argument path

mattip commented:
On windows `alloca` [is not found](https://github.com/numpy/numpy/actions/runs/5231592532/jobs/9445828884?pr=23906#step:7:335):
```
src_multiarray_methods.c.obj : error LNK2019: unresolved external symbol \
    alloca referenced in function npy_forward_method
```

It should be `_alloca` with a preceeding `_`

seberg committed:
MAINT: Just use fixed-sized array as MSVC doesn't like alloca

seberg commented:
Made it fixed-size for now...  Although I do wonder if we can rely on either `alloca` or `_alloca` for MSVC?

The reason is, that I would like to bump `NPY_MAXDIMS` _and_ `NPY_MAXARGS` (not sure how much the latter, there was once a PR for 256).  And at that point it would likely be good to optimize stack-usage a bit.

I suppose, we could still use fixed-sized arrays/buffers and make sure to only use the first part where we currently have many buffers, but I doubt that is easier to reason about (unless there are other reasons against alloca).

mattip commented:
You can use `_alloca`, you just have to guard it with a `#ifdef` macro

seberg commented:
> You can use _alloca, you just have to guard it with a #ifdef macro

Would you prefer this here right now?  I would be up to doing this especially in the places where it makes a bigger difference (ufunc machinery which does 32 args, when almost all ufuncs have 3 and iterator, but I guess also here).  After bumping up both constants.

mattip commented:
> Would you prefer this (`_alloca`) here right now?

No, I guess it is fine to do it in a future PR with benchmarks to check it is not slower.

mattip commented:
Thanks @seberg
